python/az/aro/azext_aro/custom.py

```diff
@@ -191,7 +218,66 @@ def aro_list_credentials(client, resource_group_name, resource_name):
     return client.list_credentials(resource_group_name, resource_name)
 
 
-def aro_update(client, resource_group_name, resource_name, no_wait=False):
+def aro_update(cmd, client, resource_group_name, resource_name, no_wait=False):
+    try:
+        oc = client.get(resource_group_name, resource_name)
+
+        master_subnet = oc.master_profile.subnet_id
+        worker_subnets = {w.subnet_id for w in oc.worker_profiles}
+
+        master_parts = parse_resource_id(master_subnet)
+        vnet = resource_id(
+            subscription=master_parts['subscription'],
+            resource_group=master_parts['resource_group'],
+            namespace='Microsoft.Network',
+            type='virtualNetworks',
+            name=master_parts['name'],
+        )
+
+        route_tables = get_route_tables_from_subnets(cmd.cli_ctx, worker_subnets | {master_subnet})
+
+        aad = AADManager(cmd.cli_ctx)
+        rp_client_id = os.environ.get('AZURE_FP_CLIENT_ID', FP_CLIENT_ID)
+        rp_client_sp = aad.get_service_principal(rp_client_id)
+
+        c_client_id = oc.service_principal_profile.client_id
+        c_client_sp = aad.get_service_principal(c_client_id)
+
+        if not c_client_sp:
+            logger.warning("The cluster service principal does not exist. Cannot recreate.")
+            # c_client_sp = aad.create_service_principal(c_client_id)
+
+        # Drop any None service principal objects
+        sp_obj_ids = [sp.object_id for sp in [rp_client_sp, c_client_sp] if sp]
```

Comment:
Can you explain why we have this?

Reply:
This removes any None service principal objects.  If the call to `aad.get_service_principal(client_id)` returns None meaning there isn't a service principal, we don't want to iterate over a list with a None value in it.